compiler/rustc_middle/src/ty/context/tls.rs

```diff
@@ -153,3 +154,51 @@ where
 {
     with_context_opt(|opt_context| f(opt_context.map(|context| context.tcx)))
 }
+
+/// Enters `GlobalCtxt` by setting up the `GCX_PTR` pointer.
+pub fn enter_global_context<'tcx, F, R>(context: &ImplicitCtxt<'_, 'tcx>, f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    if cfg!(parallel_compiler) && GCX_PTR.is_set() {
+        // Update `GCX_PTR` to indicate there's a `GlobalCtxt` available.
+        GCX_PTR.with(|gcx_ptr| {
+            let mut lock = gcx_ptr.value.lock();
+            assert!(lock.is_none());
+            *lock = Some(context.tcx.gcx as *const _ as *const ());
+        });
+        // Set `GCX_PTR` back to 0 when we exit.
+        let _on_drop = defer(move || {
+            GCX_PTR.with(|gcx_ptr| *gcx_ptr.value.lock() = None);
+        });
+        enter_context(context, f)
+    } else {
+        enter_context(context, f)
+    }
+}
+
+pub struct GcxPtr {
+    /// Stores a pointer to the `GlobalCtxt` if one is available.
+    /// This is used to access the `GlobalCtxt` in the deadlock handler given to Rayon.
+    value: Lock<Option<*const ()>>,
+}
+
+impl GcxPtr {
+    pub fn new() -> Self {
+        GcxPtr { value: Lock::new(None) }
+    }
+
+    /// This accesses the GlobalCtxt.
+    ///
+    /// Safety: The caller must ensure that the GlobalCtxt is live during `f`.
+    pub unsafe fn access<R>(&self, f: impl for<'tcx> FnOnce(&'tcx GlobalCtxt<'tcx>) -> R) -> R {
+        let gcx_ptr: *const GlobalCtxt<'_> = self.value.lock().unwrap() as *const _;
+        f(unsafe { &*gcx_ptr })
+    }
```

Comment:
Should we remove the `TyCtxt` inside `ImplicitCtxt`, and always use this to access it?
Can the unsafety be removing by tweaking `enter_global_context` API?

Reply:
No. Unsafety could probably be removed with some blocking scheme, but our existing scheme is simple and fast.